File: sdks/rust/src/providers/murf/socket.rs
use super::{
    protocol::{self, failure},
    settings::{self, Map, Settings},
    Input, SynthesisItem,
};
use crate::{
    generated::murf_output::*,
    http::TransportError,
    json,
    runtime::{JsonValue, StreamingInput, ValidationError},
    websocket::{Message, Socket},
};
use std::{
    collections::{BTreeMap, BTreeSet, VecDeque},
    task::{Context, Poll},
};
type Validator =
    Box<dyn Fn(&dyn std::any::Any, Option<&str>) -> Result<(), ValidationError> + Send + Sync>;
#[derive(Default)]
struct NativeContext {
    ended: bool,
    written: bool,
    flush: bool,
    audio: bool,
    final_: bool,
}
struct Write {
    fields: Map,
    end: Option<String>,
}
enum Step {
    Progress,
    Item(SynthesisItem),
}
pub(super) struct SocketStream {
    // Drop network I/O before the user producer.
    socket: Socket,
    input: StreamingInput<Input>,
    validate: Validator,
    voice: Map,
    contexts: BTreeMap<String, NativeContext>,
    order: Vec<String>,
    retired: BTreeSet<String>,
    current: Option<String>,
    session: String,
    sequence: u64,
    queue: VecDeque<Write>,
    writing: bool,
    writing_end: Option<String>,
    input_done: bool,
    limit: usize,
    prefer_output: bool,
    pub done: bool,
}
impl SocketStream {
    pub fn new(
        socket: Socket,
        mut settings: Settings,
        session: String,
        limit: usize,
        validate: Validator,
    ) -> Self {
        let queue = VecDeque::from([Write {
            fields: Map::from([
                (
                    "min_buffer_size".into(),
                    JsonValue::Number(settings.threshold),
                ),
                (
                    "max_buffer_delay_in_ms".into(),
                    JsonValue::Number(settings.delay),
                ),
            ]),
            end: None,
        }]);
        Self {
            socket,
            input: settings.input.take().unwrap(),
            validate,
            voice: settings.voice,
            contexts: BTreeMap::new(),
            order: vec![],
            retired: BTreeSet::new(),
            current: None,
            session,
            sequence: 0,
            queue,
            writing: false,
            writing_end: None,
            input_done: false,
            limit,
            prefer_output: false,
            done: false,
        }
    }
    fn end(&mut self, flush: bool) {
        if let Some(id) = self.current.take() {
            let c = self.contexts.get_mut(&id).unwrap();
            c.ended = true;
            c.flush = flush;
            self.queue.push_back(Write {
                fields: Map::from([
                    ("context_id".into(), JsonValue::String(id.clone())),
                    ("text".into(), JsonValue::String(String::new())),
                    ("end".into(), JsonValue::Bool(true)),
                ]),
                end: Some(id),
            });
        }
    }
    fn produce(&mut self, cx: &mut Context<'_>) -> Poll<Result<Step, TransportError>> {
        if self.writing {
            return Poll::Pending;
        }
        if let Some(write) = self.queue.pop_front() {
            let mut text = String::new();
            json::write(&JsonValue::Object(write.fields), &mut text)?;
            if text.len() > self.limit {
                return Poll::Ready(Err(failure("Murf message exceeds max_message_bytes")));
            }
            self.socket.as_mut().start_send(Message::Text(text))?;
            self.writing = true;
            self.writing_end = write.end;
            return Poll::Ready(Ok(Step::Progress));
        }
        if self.input_done {
            return Poll::Pending;
        }
        let item = match self.input.as_mut().poll_next(cx) {
            Poll::Pending => return Poll::Pending,
            Poll::Ready(Some(v)) => v?,
            Poll::Ready(None) => {
                self.input_done = true;
                self.end(false);
                return Poll::Ready(Ok(Step::Progress));
            }
        };
        (self.validate)(&item, Some("text"))?;
        match item {
            Input::String(text) => {
                if text.encode_utf16().count() > 3000 {
                    return Poll::Ready(Err(failure(
                        "Murf text messages must not exceed 3000 characters",
                    )));
                }
                if !text.is_empty() {
                    let id = self
                        .current
                        .get_or_insert_with(|| {
                            let id = format!("{}:{}", self.session, self.sequence);
                            self.sequence += 1;
                            self.contexts.insert(id.clone(), NativeContext::default());
                            self.order.push(id.clone());
                            id
                        })
                        .clone();
                    self.queue.push_back(Write {
                        fields: Map::from([
                            ("context_id".into(), JsonValue::String(id)),
                            ("voice_config".into(), JsonValue::Object(self.voice.clone())),
                            ("text".into(), JsonValue::String(text)),
                        ]),
                        end: None,
                    });
                }
            }
            Input::Flush(_) => self.end(true),
            Input::Clear(_) => {
                for id in self.order.drain(..) {
                    self.retired.insert(id.clone());
                    self.queue.push_back(Write {
                        fields: Map::from([
                            ("context_id".into(), JsonValue::String(id)),
                            ("clear".into(), JsonValue::Bool(true)),
                        ]),
                        end: None,
                    });
                }
                self.contexts.clear();
                self.current = None;
                // Local playback invalidation does not wait for a native clear write.
                return Poll::Ready(Ok(Step::Item(SynthesisItem::Clear(ClearEvent {
                    event: Default::default(),
                }))));
            }
            Input::Update(v) => {
                let (voice, buffering) = settings::update(v);
                self.voice.extend(voice);
                if let Some(id) = &self.current {
                    self.queue.push_back(Write {
                        fields: Map::from([
                            ("context_id".into(), JsonValue::String(id.clone())),
                            ("voice_config".into(), JsonValue::Object(self.voice.clone())),
                        ]),
                        end: None,
                    });
                }
                if !buffering.is_empty() {
                    self.queue.push_back(Write {
                        fields: buffering,
                        end: None,
                    });
                }
            }
        }
        Poll::Ready(Ok(Step::Progress))
    }
    fn receive(&mut self, cx: &mut Context<'_>) -> Poll<Result<Step, TransportError>> {
        let message = match self.socket.as_mut().poll_receive(cx) {
            Poll::Pending => return Poll::Pending,
            Poll::Ready(Some(v)) => v?,
            Poll::Ready(None) => {
                return Poll::Ready(Err(failure(
                    "Murf WebSocket closed before all contexts completed",
                )))
            }
        };
        let Message::Text(text) = message else {
            return Poll::Ready(Err(failure("Murf returned a non-text WebSocket message")));
        };
        if text.len() > self.limit {
            return Poll::Ready(Err(failure("Murf message exceeds max_message_bytes")));
        }
        let p = protocol::packet(&text)?;
        if self.retired.contains(&p.context) {
            return Poll::Ready(Ok(Step::Progress));
        }
        let c = self
            .contexts
            .get_mut(&p.context)
            .ok_or_else(|| failure("Murf returned an unknown context ID"))?;
        if !p.audio.is_empty() {
            c.audio = true;
        }
        if p.final_ {
            if !c.ended {
                return Poll::Ready(Err(failure(
                    "Murf completed a context before its text ended",
                )));
            }
            if !c.audio {
                return Poll::Ready(Err(failure("Murf completed a context without audio")));
            }
            c.final_ = true;
        }
        if p.audio.is_empty() {
            return Poll::Ready(Ok(Step::Progress));
        }
        let mut e = protocol::envelope(false);
        e.audio = Some(p.audio);
        e.correlation_id = Some(p.context);
        Poll::Ready(Ok(Step::Item(SynthesisItem::OrderedOrTimeline(e))))
    }
    pub fn next(
        &mut self,
        cx: &mut Context<'_>,
    ) -> Poll<Result<Option<SynthesisItem>, TransportError>> {
        for _ in 0..128 {
            // A failed end write must win over any native final received meanwhile.
            if self.writing {
                if let Poll::Ready(v) = self.socket.as_mut().poll_flush(cx) {
                    v?;
                    self.writing = false;
                    if let Some(id) = self.writing_end.take() {
                        if let Some(c) = self.contexts.get_mut(&id) {
                            c.written = true;
                        }
                    }
                }
            }
            if let Some(index) = self.order.iter().position(|id| {
                let c = &self.contexts[id];
                c.written && c.final_
            }) {
                let id = self.order.remove(index);
                let c = self.contexts.remove(&id).unwrap();
                if c.flush {
                    return Poll::Ready(Ok(Some(SynthesisItem::Flush(FlushEvent {
                        event: Default::default(),
                        correlation_id: id.clone(),
                        input_group_id: id,
                    }))));
                }
                continue;
            }
            if self.input_done && self.contexts.is_empty() && !self.writing && self.queue.is_empty()
            {
                self.done = true;
                return Poll::Ready(Ok(Some(SynthesisItem::Done(DoneEvent {
                    event: Default::default(),
                    remaining_characters: None,
                    warning: None,
                }))));
            }
            let first = if self.prefer_output {
                self.receive(cx)
            } else {
                self.produce(cx)
            };
            let step = match first {
                Poll::Ready(v) => {
                    self.prefer_output = !self.prefer_output;
                    v?
                }
                Poll::Pending => match if self.prefer_output {
                    self.produce(cx)
                } else {
                    self.receive(cx)
                } {
                    Poll::Pending => return Poll::Pending,
                    Poll::Ready(v) => v?,
                },
            };
            if let Step::Item(item) = step {
                return Poll::Ready(Ok(Some(item)));
            }
        }
        cx.waker().wake_by_ref();
        Poll::Pending
    }
}

File: sdks/go/providers/murf/socket.go
package murf

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	out "github.com/speechswitch/client/sdks/go/generated/murf_output"
	"github.com/speechswitch/client/sdks/go/runtime"
	"io"
	"maps"
	"sync"
	"unicode/utf16"
)

type contextState struct{ ended, written, flush, audio, final bool }
type inputResult struct {
	value Input
	err   error
}
type socketResult struct {
	value runtime.WebSocketMessage
	err   error
}
type write struct {
	value map[string]any
	endID string
}
type socketStream struct {
	parent, ctx                                      context.Context
	cancel                                           context.CancelFunc
	stop                                             func() bool
	socket                                           runtime.WebSocketLike
	input                                            runtime.Input[Input]
	voice                                            map[string]any
	threshold, delay                                 float64
	limit                                            int
	validate                                         runtime.InputValidator
	session                                          string
	sequence                                         uint64
	contexts                                         map[string]*contextState
	order                                            []string
	retired                                          map[string]bool
	current                                          string
	once                                             sync.Once
	mutex, lifecycle                                 sync.Mutex
	closeError                                       error
	inputIdle                                        <-chan struct{}
	started, terminal, done, inputDone, preferOutput bool
	pendingInput                                     <-chan inputResult
	pendingOutput                                    <-chan socketResult
	pendingSend                                      <-chan error
	writingEnd                                       string
	queue                                            []write
}

func (s *socketStream) closeResources() error {
	s.once.Do(func() {
		s.lifecycle.Lock()
		s.cancel()
		idle := s.inputIdle
		s.lifecycle.Unlock()
		s.closeError = s.socket.Close()
		// A blocked user producer must not keep the transport open or race its Close.
		go func() { <-idle; _ = s.input.Close() }()
	})
	return s.closeError
}
func (s *socketStream) Close() error { s.stop(); return s.closeResources() }
func (s *socketStream) Next(ctx context.Context) (out.SynthesisItem, error) {
	stop := context.AfterFunc(ctx, func() { s.closeResources() })
	defer stop()
	s.mutex.Lock()
	defer s.mutex.Unlock()
	if s.terminal || s.done {
		return nil, io.EOF
	}
	item, err := s.next(ctx)
	if cause := ctx.Err(); cause != nil {
		item, err = nil, cause
	} else if cause := s.parent.Err(); cause != nil {
		item, err = nil, cause
	} else if s.ctx.Err() != nil && !s.done {
		item, err = nil, io.EOF
	}
	if err != nil {
		s.terminal = true
		s.Close()
	}
	return item, err
}
func (s *socketStream) pullInput() {
	s.lifecycle.Lock()
	defer s.lifecycle.Unlock()
	if s.ctx.Err() != nil {
		return
	}
	pending, idle := make(chan inputResult, 1), make(chan struct{})
	s.pendingInput, s.inputIdle = pending, idle
	go func() { value, err := s.input.Next(s.ctx); close(idle); pending <- inputResult{value, err} }()
}
func (s *socketStream) pullOutput() {
	pending := make(chan socketResult, 1)
	s.pendingOutput = pending
	go func() { value, err := s.socket.Receive(s.ctx); pending <- socketResult{value, err} }()
}
func (s *socketStream) send(w write) {
	pending := make(chan error, 1)
	s.pendingSend, s.writingEnd = pending, w.endID
	// Snapshot before starting I/O; only Next owns mutable protocol state.
	data, err := json.Marshal(w.value)
	if err == nil && len(data) > s.limit {
		err = errors.New("Murf message exceeds MaxMessageBytes")
	}
	go func() {
		if err == nil {
			err = s.socket.Send(s.ctx, runtime.WebSocketText(data))
		}
		pending <- err
	}()
}
func (s *socketStream) end(flush bool) {
	if s.current == "" {
		return
	}
	id := s.current
	s.current = ""
	c := s.contexts[id]
	c.ended, c.flush = true, flush
	s.queue = append(s.queue, write{map[string]any{"context_id": id, "text": "", "end": true}, id})
}
func (s *socketStream) takeInput(value Input) (out.SynthesisItem, error) {
	if err := s.validate(value); err != nil {
		return nil, err
	}
	kind, text, voice, buffering := command(value)
	switch kind {
	case "text":
		if len(utf16.Encode([]rune(text))) > 3000 {
			return nil, errors.New("Murf text messages must not exceed 3000 characters")
		}
		if text != "" {
			if s.current == "" {
				s.current = fmt.Sprintf("%s:%d", s.session, s.sequence)
				s.sequence++
				s.contexts[s.current] = &contextState{}
				s.order = append(s.order, s.current)
			}
			s.queue = append(s.queue, write{value: map[string]any{"context_id": s.current, "voice_config": maps.Clone(s.voice), "text": text}})
		}
	case "flush":
		s.end(true)
	case "clear":
		for _, id := range s.order {
			s.retired[id] = true
			s.queue = append(s.queue, write{value: map[string]any{"context_id": id, "clear": true}})
		}
		clear(s.contexts)
		s.order = nil
		s.current = ""
		// This is local playback invalidation, emitted before potentially stalled writes.
		return out.SynthesisItemAsClear{}, nil
	case "update":
		maps.Copy(s.voice, voice)
		if s.current != "" {
			s.queue = append(s.queue, write{value: map[string]any{"context_id": s.current, "voice_config": maps.Clone(s.voice)}})
		}
		if len(buffering) > 0 {
			s.queue = append(s.queue, write{value: buffering})
		}
	}
	return nil, nil
}

func (s *socketStream) next(ctx context.Context) (out.SynthesisItem, error) {
	for {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		if s.ctx.Err() != nil {
			return nil, io.EOF
		}
		if !s.started {
			s.started = true
			s.send(write{value: map[string]any{"min_buffer_size": s.threshold, "max_buffer_delay_in_ms": s.delay}})
			s.pullOutput()
		}
		for index, id := range s.order {
			c := s.contexts[id]
			if c.final && c.written {
				delete(s.contexts, id)
				s.order = append(s.order[:index], s.order[index+1:]...)
				if c.flush {
					return out.SynthesisItemAsFlush{Value: out.FlushEvent{CorrelationId: id, InputGroupId: id}}, nil
				}
				break
			}
		}
		if s.pendingSend == nil {
			if len(s.queue) > 0 {
				w := s.queue[0]
				s.queue[0] = write{}
				s.queue = s.queue[1:]
				s.send(w)
			} else if s.inputDone && len(s.contexts) == 0 {
				s.done = true
				s.Close()
				return out.SynthesisItemAsDone{}, nil
			} else if !s.inputDone && s.pendingInput == nil {
				s.pullInput()
			}
		}
		var kind string
		var input inputResult
		var output socketResult
		var sendError error
		select {
		case sendError = <-s.pendingSend:
			kind = "send"
		default:
		}
		if kind == "" {
			if s.preferOutput {
				select {
				case output = <-s.pendingOutput:
					kind = "output"
				default:
					select {
					case input = <-s.pendingInput:
						kind = "input"
					default:
					}
				}
			} else {
				select {
				case input = <-s.pendingInput:
					kind = "input"
				default:
					select {
					case output = <-s.pendingOutput:
						kind = "output"
					default:
					}
				}
			}
		}
		if kind == "" {
			select {
			case <-ctx.Done():
				return nil, ctx.Err()
			case <-s.ctx.Done():
				return nil, io.EOF
			case sendError = <-s.pendingSend:
				kind = "send"
			case input = <-s.pendingInput:
				kind = "input"
			case output = <-s.pendingOutput:
				kind = "output"
			}
		}
		switch kind {
		case "send":
			s.pendingSend = nil
			if sendError != nil {
				return nil, sendError
			}
			if c := s.contexts[s.writingEnd]; c != nil {
				c.written = true
			}
			s.writingEnd = ""
		case "input":
			s.pendingInput = nil
			s.preferOutput = true
			if input.err == io.EOF {
				s.inputDone = true
				s.end(false)
				continue
			}
			if input.err != nil {
				return nil, input.err
			}
			item, err := s.takeInput(input.value)
			if item != nil || err != nil {
				return item, err
			}
		case "output":
			s.pendingOutput = nil
			s.preferOutput = false
			if output.err == io.EOF {
				return nil, errors.New("Murf WebSocket closed before all contexts completed")
			}
			if output.err != nil {
				return nil, output.err
			}
			frame, ok := output.value.(runtime.WebSocketText)
			if !ok {
				return nil, errors.New("Murf returned a non-text WebSocket message")
			}
			if len(frame) > s.limit {
				return nil, errors.New("Murf message exceeds MaxMessageBytes")
			}
			p, err := decodePacket([]byte(frame))
			if err != nil {
				return nil, err
			}
			s.pullOutput()
			if s.retired[p.context] {
				continue
			}
			c := s.contexts[p.context]
			if c == nil {
				return nil, errors.New("Murf returned an unknown context ID")
			}
			if len(p.audio) > 0 {
				c.audio = true
			}
			if p.final {
				if !c.ended {
					return nil, errors.New("Murf completed a context before its text ended")
				}
				if !c.audio {
					return nil, errors.New("Murf completed a context without audio")
				}
				c.final = true
			}
			if len(p.audio) > 0 {
				return out.SynthesisItemAsOrderedOrTimeline{Value: out.MurfEnvelope{Correlation: out.MurfEnvelopeCorrelationAsOrdered{}, CorrelationId: runtime.Some(p.context), Audio: runtime.Some(p.audio), Timestamps: []out.MurfTimestamp{}}}, nil
			}
		}
	}
}
